File: homework/task-2/code/container.cpp
//------------------------------------------------------------------------------
// container.cpp - содержит функции обработки контейнера
//------------------------------------------------------------------------------

#include "container.h"

#include <sstream>

//------------------------------------------------------------------------------
// Очистка контейнера от элементов (освобождение памяти)
container::~container() {
    for (int i = 0; i < len; i++) {
        delete cont[i];
    }
    len = 0;
}

container::container(container&& c) {
    len = c.len;
    for (int i = 0; i < len; i++) {
        cont[i] = c.cont[i];
    }
    c.len = 0;
}

container& container::operator=(container&& c) {
    this->~container();

    len = c.len;
    for (int i = 0; i < len; i++) {
        cont[i] = c.cont[i];
    }
    c.len = 0;

    return *this;
}

//------------------------------------------------------------------------------
// Ввод содержимого контейнера из указанного потока
container::container(std::ifstream& ifst) {
    while (!ifst.eof()) {
        if ((cont[len] = plant::create(ifst)) != nullptr) {
            len++;
        }
    }
}

//------------------------------------------------------------------------------
// Случайный ввод содержимого контейнера
container::container(const Random& rand, int size) {
    while (len < size) {
        if ((cont[len] = plant::createRnd(rand)) != nullptr) {
            len++;
        }
    }
}

//------------------------------------------------------------------------------
// Вывод содержимого контейнера в указанный поток
std::string container::toString() const {
    std::stringstream osst;
    osst << "Container contains " << len << " elements.\n";
    for (int i = 0; i < len; i++) {
        osst << i << ": ";
        osst << cont[i]->toString();
    }
    return osst.str();
}

// Сортировка контейнера методом Straight Selection
void container::sort() {
    int i = 0, j, k;
    plant *temp;
    while (i < len) {
        k = i;
        for (j = i + 1; j < len; j++) {
            if (cont[j]->Hash() < cont[k]->Hash()) {
                k = j;
            }
        }
        temp = cont[i];
        cont[i] = cont[k];
        cont[k] = temp;
        i++;
    }
}

File: homework/task-3/src/container.py
# container.py - содержит класс контейнера

from io import TextIOWrapper
from typing import Union

import random

from plant import Plant


class Container:
    def __init__(self, x: Union[TextIOWrapper, random.Random], size=0):
        self.cont = []
        if isinstance(x, TextIOWrapper):
            self.init_from_file(x)
        else:
            self.init_from_random(x, size)

    # Ввод содержимого контейнера из файла
    def init_from_file(self, f: TextIOWrapper):
        for line in f.readlines():
            self.cont.append(Plant.create_from_str(line))

    # Случайный ввод содержимого контейнера
    def init_from_random(self, r: random.Random, size=0):
        for i in range(size):
            self.cont.append(Plant.create_from_rand(r))

    # Вывод содержимого контейнера в строку
    def __str__(self):
        s = ''
        s += "Container contains " + str(len(self.cont)) + " elements.\n"
        for i in range(len(self.cont)):
            s += str(i) + ": " + str(self.cont[i])
        return s

    # Сортировка контейнера методом Straight Selection
    def sort(self):
        i = 0
        while i < len(self.cont):
            k = i
            j = i + 1
            while j < len(self.cont):
                if self.cont[j].hash() < self.cont[k].hash():
                    k = j
                j += 1
            temp = self.cont[i]
            self.cont[i] = self.cont[k]
            self.cont[k] = temp
            i += 1
